samples/drivers/i2s/src/main.c

```diff
@@ -0,0 +1,207 @@
+	i2s_cfg.format = I2S_FMT_DATA_FORMAT_I2S;
+	i2s_cfg.options = I2S_OPT_LOOPBACK;
+	i2s_cfg.frame_clk_freq = 8000;
+	i2s_cfg.block_size = BLOCK_SIZE;
+	i2s_cfg.mem_slab = &rx_mem_slab;
+	i2s_cfg.timeout = K_FOREVER;
+
+	ret = i2s_configure(dev_i2s, I2S_DIR_RX, &i2s_cfg);
+	if (ret < 0) {
+		printk("Error: Failed to configure I2S RX stream\n");
+		return;
+	}
+
+	i2s_cfg.mem_slab = &tx_mem_slab;
+
+	ret = i2s_configure(dev_i2s, I2S_DIR_TX, &i2s_cfg);
+	if (ret < 0) {
+		printk("Error: Failed to configure I2S TX stream\n");
+		return;
+	}
+
+	/*
+	 * Transmit / Receive data
+	 */
+
+	printk("Start I2S bus transmission\n");
+
+	ret = k_mem_slab_alloc(&tx_mem_slab, &tx_block, K_FOREVER);
+	if (ret < 0) {
+		printk("Error: Failed to allocate tx_block\n");
+		return;
+	}
+	fill_buf((u16_t *)tx_block, 0);
+	ret = i2s_write(dev_i2s, tx_block, BLOCK_SIZE);
+	if (ret < 0) {
+		printk("Error: Write failed\n");
+		return;
+	}
+	printk("1->OK\n");
+
+	ret = k_mem_slab_alloc(&tx_mem_slab, &tx_block, K_FOREVER);
+	if (ret < 0) {
+		printk("Error: Failed to allocate tx_block\n");
+		return;
+	}
+	fill_buf((u16_t *)tx_block, 1);
+	ret = i2s_write(dev_i2s, tx_block, BLOCK_SIZE);
+	if (ret < 0) {
+		printk("Error: Write failed\n");
+		return;
+	}
+	printk("2->OK\n");
+
+	/* Start transmission */
+	ret = i2s_trigger(dev_i2s, I2S_DIR_TX | I2S_DIR_RX, I2S_TRIGGER_START);
+	if (ret < 0) {
+		printk("Error: Trigger failed\n");
+		return;
+	}
+
+	ret = i2s_read(dev_i2s, &rx_block, &rx_size);
+	if (ret < 0 || rx_size != BLOCK_SIZE) {
+		printk("Error: Read failed\n");
+		return;
+	}
+	ret = verify_buf((u16_t *)rx_block, 0);
+	if (ret < 0) {
+		printk("Error: Verify failed\n");
+		return;
+	}
+	printk("1<-OK\n");
+	k_mem_slab_free(&rx_mem_slab, &rx_block);
+
+	ret = k_mem_slab_alloc(&tx_mem_slab, &tx_block, K_FOREVER);
+	if (ret < 0) {
+		printk("Error: Failed to allocate tx_block\n");
+		return;
+	}
+	fill_buf((u16_t *)tx_block, 2);
+	ret = i2s_write(dev_i2s, tx_block, BLOCK_SIZE);
+	if (ret < 0) {
+		printk("Error: Write failed\n");
+		return;
+	}
+	printk("3->OK\n");
+
+	/* All data written, flush the queue and stop the transmission */
+	ret = i2s_trigger(dev_i2s, I2S_DIR_TX | I2S_DIR_RX, I2S_TRIGGER_FLUSH);
+	if (ret < 0) {
+		printk("Error: Trigger failed\n");
+		return;
+	}
+
+	ret = i2s_read(dev_i2s, &rx_block, &rx_size);
+	if (ret < 0 || rx_size != BLOCK_SIZE) {
+		printk("Error: Read failed\n");
+		return;
+	}
+	ret = verify_buf((u16_t *)rx_block, 1);
+	if (ret < 0) {
+		printk("Error: Verify failed\n");
+		return;
+	}
+	printk("2<-OK\n");
+	k_mem_slab_free(&rx_mem_slab, &rx_block);
+
+	ret = i2s_read(dev_i2s, &rx_block, &rx_size);
+	if (ret < 0 || rx_size != BLOCK_SIZE) {
+		printk("Error: Read failed\n");
+		return;
+	}
+	ret = verify_buf((u16_t *)rx_block, 2);
+	if (ret < 0) {
+		printk("Error: Verify failed\n");
+		return;
+	}
+	printk("3<-OK\n");
+	k_mem_slab_free(&rx_mem_slab, &rx_block);
+
+	printk("I2S Transmission finished successfully\n");
+}
```

Comment:
At some point we should have an audio subsystem. I believe though that Zephyr, being an embedded OS, should expose low level drivers like I2S since some users may have special needs / will prefer not to use higher level APIs.

If this sample app is too complex - it seems like it is - I may simplify it. In my opinion we will need another much more sophisticated app to be used as a test.